File: Kalk/Logica/rettangolo.cpp
#include "rettangolo.h"

Rettangolo::Rettangolo(const Punto& a,const Punto& b,const Punto& c,const Punto& d):Spezzata(a,a){
    Spezzata::aggiungiPunto(b);
    Spezzata::aggiungiPunto(c);
    Spezzata::aggiungiPunto(d);
}

Rettangolo::Rettangolo(const Punto & a):Spezzata(a,a){}



Tag* Rettangolo::operator+(const Tag& t) const{
    return Spezzata::operator +(t);
}

QString Rettangolo::output()const{
    QString stringaTag,stringaCoord;
    stringaCoord=stringaCoord+" ("+QString::number(getInizio().x())+","+
            QString::number(getInizio().y())+"),";
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        stringaCoord=stringaCoord+"("+QString::number(cit->x())+","+
                QString::number(cit->y())+"),";
    }
    return getTag()+stringaCoord.remove(stringaCoord.length()-1,1);
}

Rettangolo* Rettangolo::simmetricoX(QString etichetta)const{
    Spezzata* s=Spezzata::simmetricoX(etichetta);
    Rettangolo* r=new Rettangolo(s->getInizio());
    for(auto cit=s->punti.constBegin();cit!=s->punti.constEnd();++cit){
        r->Spezzata::aggiungiPunto(*(cit));
    }
    return r;
}
Rettangolo* Rettangolo::simmetricoY(QString etichetta)const{
    Spezzata* s=Spezzata::simmetricoY(etichetta);
    Rettangolo* r=new Rettangolo(s->getInizio());
    for(auto cit=s->punti.constBegin();cit!=s->punti.constEnd();++cit){
        r->Spezzata::aggiungiPunto(*(cit));
    }
    return r;

}
Rettangolo* Rettangolo::simmetricoO(QString etichetta)const{
    Spezzata* s=Spezzata::simmetricoO(etichetta);
    Rettangolo* r=new Rettangolo(s->getInizio());
    for(auto cit=s->punti.constBegin();cit!=s->punti.constEnd();++cit){
        r->Spezzata::aggiungiPunto(*(cit));
    }
    return r;
}

void Rettangolo::traslazione(double nX,double nY){
    const_cast<Punto&>(getInizio()).traslazione(nX,nY);
    for(auto it=punti.begin();it!=punti.end();++it){
        it->traslazione(nX,nY);
    }
}

double Rettangolo::lunghezza()const{
    return Spezzata::lunghezza();
}

double Rettangolo::area()const{
    return getInizio().distanzaP(punti.first()) * getInizio().distanzaP(punti.last());
}

File: java/Kalk/src/Kalk/Classes/Rettangolo.java
package Kalk.Classes;

import java.util.ListIterator;

public class Rettangolo extends Spezzata {

	public Rettangolo(String s, Punto a, Punto b, Punto c, Punto d) {
		super(s, a, a);
		aggiungi(b);
		aggiungi(c);
		aggiungi(d);
	}
	
	public String toString() {
		String c="";
		c=c+" ("+getInizio().x()+","+getInizio().y()+"),";
		ListIterator<Punto> i;
		i=getLista().listIterator();
		while(i.hasNext()) {
			c=c+"("+i.next().x();
			i.previous();
			c=c+","+i.next().y()+"),";
		}
		return getTag()+c;
	}
	
	public double lunghezza() {
		return (getInizio().distanzaP(getLista().get(0)) + getInizio().distanzaP(getLista().get(2)))*2;
	}
	
	public double area() {
		return getInizio().distanzaP(getLista().get(0)) * getInizio().distanzaP(getLista().get(2));
	}
	
	public Rettangolo simmetricoX(String s) {
		Spezzata sp=super.simmetricoX(s);
		return new Rettangolo(getTag(),sp.getInizio(),sp.getLista().get(0),sp.getLista().get(1),sp.getLista().get(2));
	}
	
	public Rettangolo simmetricoY(String s) {
		Spezzata sp=super.simmetricoY(s);
		return new Rettangolo(getTag(),sp.getInizio(),sp.getLista().get(0),sp.getLista().get(1),sp.getLista().get(2));
	}
	
	public Rettangolo simmetricoO(String s) {
		Spezzata sp=super.simmetricoO(s);
		return new Rettangolo(getTag(),sp.getInizio(),sp.getLista().get(0),sp.getLista().get(1),sp.getLista().get(2));
	}
	
	public void traslazione(double nX, double nY) {
		getInizio().traslazione(nX, nY);
		getLista().forEach((Punto)->{Punto.traslazione(nX, nY);});
	}
}
